presto-main/src/test/java/io/prestosql/server/security/oauth2/TestingHydraService.java

```diff
@@ -0,0 +1,183 @@
+        closer.register(databaseContainer);
+        closer.register(migrationContainer);
+        closer.register(consentContainer);
+        closer.register(hydraContainer);
+    }
+
+    public void start()
+    {
+        databaseContainer.start();
+        migrationContainer.start();
+        consentContainer.start();
+        hydraContainer.start();
+    }
+
+    public FixedHostPortGenericContainer<?> createHydraContainer()
+    {
+        return new FixedHostPortGenericContainer<>(HYDRA_IMAGE).withNetwork(network);
+    }
+
+    public Network getNetwork()
+    {
+        return network;
+    }
+
+    public int getHydraPort()
+    {
+        return hydraContainer.getMappedPort(4444);
+    }
+
+    @Override
+    public void close()
+            throws IOException
+    {
+        try {
+            closer.close();
+        }
+        catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static String generateSecret()
+    {
+        byte[] randomBytes = new byte[32];
+        SECURE_RANDOM.nextBytes(randomBytes);
+        return Base64.getEncoder().encodeToString(randomBytes);
+    }
+
+    public static void main(String[] args)
+            throws Exception
+    {
+        try (TestingHydraService service = new TestingHydraService()) {
+            // expose containers ports & override environment variables
+            service.consentContainer
+                    .withFixedExposedPort(9020, 3000);
+            service.hydraContainer
+                    .withFixedExposedPort(9001, 4444)
+                    .withFixedExposedPort(9002, 4445)
+                    .withEnv("URLS_SELF_ISSUER", "http://localhost:9001/")
+                    .withEnv("URLS_CONSENT", "http://localhost:9020/consent")
+                    .withEnv("URLS_LOGIN", "http://localhost:9020/login")
+                    .withEnv("TTL_ACCESS_TOKEN", "30m");
+            service.start();
+
+            // create a client for Presto in Hydra
+            service.createHydraContainer()
+                    .withCommand("clients create " +
+                            "--endpoint http://hydra:4445 " +
+                            "--id another-consumer " +
+                            "--secret consumer-secret " +
+                            "-g authorization_code,refresh_token " +
+                            "-r token,code,id_token " +
+                            "--scope openid,offline " +
+                            "--callbacks https://localhost:8443/oauth2/callback")
+                    .withStartupCheckStrategy(new OneShotStartupCheckStrategy().withTimeout(Duration.ofSeconds(30)))
+                    .start();
+            try (TestingPrestoServer ignored = TestingPrestoServer.builder()
+                    .setCoordinator(true)
+                    .setAdditionalModule(new WebUiModule())
+                    .setProperties(
+                            ImmutableMap.<String, String>builder()
```

Comment:
This code is also duplicated in `TestOAuth2WebUiAuthenticationFilter`. Can we move it somewhere common?

While duplication is not always a problem, since this is a `main()` method, it's not tested, so having it use tested code is better.

Reply:
It's not so straight forward as in the previous case and there are quite a few differences between those two places and I would prefer to leave it as it is.